script-catboost.py

```diff
@@ -0,0 +1,459 @@
+import matplotlib.pyplot as plt
+from sklearn.model_selection import TimeSeriesSplit
+from sklearn.metrics import root_mean_squared_error, log_loss, roc_auc_score
+from statsmodels.nonparametric.smoothers_lowess import lowess
+from concurrent.futures import ProcessPoolExecutor, as_completed
+from itertools import accumulate
+import torch
+from catboost import CatBoostClassifier, Pool
+
+dev_mode = os.environ.get("DEV_MODE")
+
+if dev_mode:
+    # for local development
+    sys.path.insert(0, os.path.abspath("../fsrs-optimizer/src/fsrs_optimizer/"))
+
+from fsrs_optimizer import (
+    Optimizer,
+    Trainer,
+    FSRS,
+    Collection,
+    power_forgetting_curve,
+    remove_outliers,
+    remove_non_continuous_rows,
+    plot_brier,
+    rmse_matrix,
+)
+
+
+model = FSRS
+optimizer = Optimizer()
+lr: float = 4e-2
+n_epoch: int = 5
+n_splits: int = 5
+batch_size: int = 512
+verbose: bool = False
+verbose_inadequate_data: bool = False
+do_fullinfo_stats: bool = False
+
+dry_run = os.environ.get("DRY_RUN")
+only_pretrain = os.environ.get("PRETRAIN")
+rust = os.environ.get("FSRS_RS")
+if rust:
+    path = "FSRS-rs-catboost"
+    if do_fullinfo_stats:
+        path += "-fullinfo"
+    from anki._backend import RustBackend
+
+    backend = RustBackend()
+
+else:
+    path = "FSRS-5-catboost"
+    if dry_run:
+        path += "-dry-run"
+    if only_pretrain:
+        path += "-pretrain"
+    if dev_mode:
+        path += "-dev"
+    if do_fullinfo_stats:
+        path += "-fullinfo"
+
+
+def predict(w_list, trainsets, testsets, file=None):
+    p = []
+    y = []
+    save_tmp = [] if file else None
+
+    for i, (w, trainset, testset) in enumerate(zip(w_list, trainsets, testsets)):
+        my_collection = Collection(w)
+        trainset["stability"], trainset["difficulty"] = my_collection.batch_predict(
+            trainset
+        )
+        trainset["pf"] = power_forgetting_curve(
+            trainset["delta_t"], trainset["stability"]
+        )
+
+        testset["stability"], testset["difficulty"] = my_collection.batch_predict(
+            testset
+        )
+        testset["pf"] = power_forgetting_curve(testset["delta_t"], testset["stability"])
+
+        catboost_traindata = trainset[["delta_t", "pf", "stability", "difficulty"]]
+        train_pool = Pool(
+            catboost_traindata,
+            label=trainset["y"].astype(int),
+        )
+        catboost_evaldata = testset[["delta_t", "pf", "stability", "difficulty"]]
+        eval_pool = Pool(
+            catboost_evaldata,
+            label=testset["y"].astype(int),
+        )
+
+        model = CatBoostClassifier(
+            verbose=False, allow_writing_files=False, has_time=True
+        )
+        f = model.fit(train_pool, eval_set=eval_pool, early_stopping_rounds=100)
+        # print(f"Model accuracy on eval data: {f.score(eval_pool):.5%}")
+        # print(f.get_feature_importance(prettified=True))
+
+        prob = f.predict_proba(eval_pool)
```

Comment:
It's not very useful for production if we predict the probability directly because the final result is the interval scheduled for the user. How can we find out the inverse function of it?

Reply:
For predicting interval you would need to build a completely different booster. You would swap delta_t and y.
Then probably would be good idea to change y to rating if you want all difficulty buttons and maybe delta_t for the last correct delta_t.

Reply:
For predicting interval you would need to build a completely different booster. You would swap delta_t and y.
Then probably would be good idea to change y to rating if you want all difficulty buttons and maybe delta_t for the last correct delta_t.